CSC8503CoreClasses/RecastBuilder.cpp

```diff
@@ -0,0 +1,301 @@
+}
+
+RecastBuilder::~RecastBuilder() {
+	cleanup();
+}
+
+void RecastBuilder::BuildNavMesh(std::vector<GameObject*> objects) {
+	if (objects.empty()) return;
+
+	cleanup();
+
+	// Translating the meshes into useable values for Recast
+	float* bmin = new float[3] {std::numeric_limits<float>::max(), std::numeric_limits<float>::max(), std::numeric_limits<float>::max()};
+	float* bmax = new float[3] {-std::numeric_limits<float>::max(), -std::numeric_limits<float>::max(), -std::numeric_limits<float>::max()};
+	int currentTriCount = 0;
+	int vertCount = 0;
+	int trisCount = 0;
+	for (int i = 0; i < objects.size(); i++) {
+		vertCount += objects[i]->GetRenderObject()->GetMesh()->GetVertexCount();
+		trisCount += objects[i]->GetRenderObject()->GetMesh()->GetPrimitiveCount();
+	}
+	float* verts = new float[vertCount * 3];
+	unsigned int* tris = new unsigned int[trisCount * 3];
+	int vCount = 0;
+	int tCount = 0;
+	vertCount = 0;
+	for (int i = 0; i < objects.size(); i++) {
+		int lastVertCount = vertCount;
+		std::vector<Vector3> vertsVector = objects[i]->GetRenderObject()->GetMesh()->GetPositionData();
+		for (int j = 0; j < vertsVector.size(); j++) {
+			verts[vCount++] = (vertsVector[j].x * objects[i]->GetTransform().GetScale().x) + objects[i]->GetTransform().GetPosition().x;
+			verts[vCount++] = (vertsVector[j].y * objects[i]->GetTransform().GetScale().y) + objects[i]->GetTransform().GetPosition().y;
+			verts[vCount++] = (vertsVector[j].z * objects[i]->GetTransform().GetScale().z) + objects[i]->GetTransform().GetPosition().z;
+			bmin[0] = std::min(bmin[0], verts[vCount - 3]);
+			bmin[1] = std::min(bmin[1], verts[vCount - 2]);
+			bmin[2] = std::min(bmin[2], verts[vCount - 1]);
+			bmax[0] = std::max(bmax[0], verts[vCount - 3]);
+			bmax[1] = std::max(bmax[1], verts[vCount - 2]);
+			bmax[2] = std::max(bmax[2], verts[vCount - 1]);
+		}
+
+		std::vector<unsigned int> trisVector = objects[i]->GetRenderObject()->GetMesh()->GetIndexData();
+		for (int j = 0; j < trisVector.size(); j++) {
+			tris[tCount] = trisVector[j] + lastVertCount;
+			tCount++;
+		}
+		vertCount += objects[i]->GetRenderObject()->GetMesh()->GetVertexCount();
+	}
+	if (!InitialiseConfig(bmin, bmax)) return;
+	if (!RasterizeInputPolygon(verts, vertCount, tris, trisCount)) return;
+	if (!FilterWalkableSurfaces()) return;
+	if (!PartitionWalkableSurface()) return;
+	if (!TraceContours()) return;
+	if (!BuildPoly()) return;
+	if (!BuildDetailPoly()) return;
+	if (!CreateDetourData()) return;
+}
+
+bool RecastBuilder::InitialiseConfig(const float* bmin, const float* bmax) {
+	memset(&mConfig, 0, sizeof(mConfig));
+
+	mConfig.cs = mCellSize;
+	mConfig.ch = mCellHeight;
+	mConfig.walkableSlopeAngle = mGuardMaxSlope;
+	mConfig.walkableHeight = (int)ceilf(mGuardHeight / mConfig.ch);
+	mConfig.walkableClimb = (int)floorf(mGuardMaxClimb / mConfig.ch);
+	mConfig.walkableRadius = (int)ceilf(mGuardRadius / mConfig.cs);
+	mConfig.maxEdgeLen = (int)(mMaxEdgeLength / mCellSize);
```

Comment:
When doing a division, it's best to add a check a dividing number is not a zero for preventing the runtime error ✌️